mito-ai/src/Extensions/InlineCompleter/provider.ts

```diff
@@ -0,0 +1,403 @@
+import { Notification, showErrorMessage } from '@jupyterlab/apputils';
+import { IEditorMimeTypeService } from '@jupyterlab/codeeditor';
+import {
+  IEditorLanguageRegistry,
+  type IEditorLanguage
+} from '@jupyterlab/codemirror';
+import {
+  InlineCompletionTriggerKind,
+  type CompletionHandler,
+  type IInlineCompletionContext,
+  type IInlineCompletionItem,
+  type IInlineCompletionList,
+  type IInlineCompletionProvider
+} from '@jupyterlab/completer';
+import { DocumentWidget } from '@jupyterlab/docregistry';
+import { NotebookPanel } from '@jupyterlab/notebook';
+import type { ISettingRegistry } from '@jupyterlab/settingregistry';
+import { PromiseDelegate, type JSONValue } from '@lumino/coreutils';
+import type { IDisposable } from '@lumino/disposable';
+import { Signal, Stream } from '@lumino/signaling';
+import {
+  CompletionWebsocketClient,
+  type ICompletionWebsocketClientOptions
+} from './client';
+import type { CompletionError, InlineCompletionStreamChunk } from './models';
+
+/**
+ * Mito AI inline completer
+ *
+ * It uses a WebSocket connection to request an AI model.
+ */
+export class MitoAIInlineCompleter
+  implements IInlineCompletionProvider, IDisposable
+{
+  private _client: CompletionWebsocketClient;
+  private _counter = 0;
+  private _isDisposed = false;
+  private _languageRegistry: IEditorLanguageRegistry;
+  private _settings: MitoAIInlineCompleter.ISettings =
+    MitoAIInlineCompleter.DEFAULT_SETTINGS;
+  // Store only one inline completion stream
+  //   Each new request should invalidate any other suggestions.
+  private _currentPrefix = '';
+  private _currentToken = '';
+  private _currentStream: Stream<
+    MitoAIInlineCompleter,
+    InlineCompletionStreamChunk
+  > | null = null;
+  // Block processing chunks while waiting for the acknowledge request
+  // that will provide the unique completion token;
+  private _completionLock = new PromiseDelegate<void>();
+  private _fullCompletionMap = new WeakMap<
+    Stream<MitoAIInlineCompleter, InlineCompletionStreamChunk>,
+    string
+  >();
+
+  constructor({
+    languageRegistry,
+    ...clientOptions
+  }: MitoAIInlineCompleter.IOptions) {
+    this._languageRegistry = languageRegistry;
+    this._client = new CompletionWebsocketClient(clientOptions);
+
+    this._client
+      .initialize()
+      .then(() => {
```

Comment:
This could be rewritten using the trick of defining a local async function if you want to avoid this promise chain in the constructor.